Ensure IOError, and failed unit test exit code is '1'

We need to ensure that an [`IOError`](https://travis-ci.org/jeff1evesque/machine-learning/builds/150791961#L2977-L2981) generated from `factory.py` causes the corresponding pytest to have an exit code of `1` within the travis ci build.  Most importantly, we need to make sure that unit tests that fail, generate corresponding exit code value of `1`, within the corresponding docker container, executed via the travis ci build.

One possible approach, is to determine if we can redirect the output of an [executed command](https://travis-ci.org/jeff1evesque/machine-learning/builds/151104916#L3158-L3231), in the `.travis.yaml`, to a variable.  Then, within the variable, we could `grep` for the following line:

```bash
=================================== FAILURES ===================================
```

If such a line exists, we could `return 1` in the `.travis.yml`, to simulate a failed travis ci build.

One possible approach, is to determine if we can pipe the output of an [executed command](https://travis-ci.org/jeff1evesque/machine-learning/builds/151104916#L3158-L3231), in the `.travis.yaml`, to a `grep`, for the following line:

```bash
=================================== FAILURES ===================================
```

If such a line exists, we could `return 1` in the `.travis.yml`, to simulate a failed travis ci build.

We've attempted to return a non-zero exit code, from python logic inside the `webserver` docker container instance, to the exit code of the corresponding docker container.  However, all attempts have failed.  Another possible approach, is to determine if we can pipe the output of an [executed command](https://travis-ci.org/jeff1evesque/machine-learning/builds/151104916#L3158-L3231), in the `.travis.yaml`, to a `grep`, for the following line:

```bash
=================================== FAILURES ===================================
```

If such a line exists, we could `return 1` in the `.travis.yml`, to simulate a failed travis ci build.

We've attempted to `return` a non-zero exit code, from python logic inside the `webserver` docker container instance, to the exit code of the corresponding docker container.  However, all attempts have failed.  Another possible approach, is to determine if we can pipe the output of an [executed command](https://travis-ci.org/jeff1evesque/machine-learning/builds/151104916#L3158-L3231) (in the `.travis.yaml`), to a `grep`, for the following line:

```bash
=================================== FAILURES ===================================
```

If such a line exists, we could `return 1` in the `.travis.yml`, to simulate a failed travis ci build.

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,12 +105,12 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && \
-    docker run --name base --net=app_nw -d container-default && \
-    docker run --name redis --net=app_nw -d container-redis && \
-    docker run --name webserver --net=app_nw -d container-webserver run && \
-    docker run --name database --net=app_nw -d container-database && \
-    docker run --name webserver-pytest --net=app_nw -it container-webserver test | \
+  - docker network create -d bridge app_nw \
+    && docker run --name base --net=app_nw -d container-default \
+    && docker run --name redis --net=app_nw -d container-redis \
+    && docker run --name webserver --net=app_nw -d container-webserver run \
+    && docker run --name database --net=app_nw -d container-database \
+    && docker run --name webserver-pytest --net=app_nw -it container-webserver test | \
     grep '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -q '= FAILURES =' && return 1
+  - cat pytest.log | grep -q '= FAILURES ='
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - grep -c '= FAILURES =' pytest.log
+  - grep -c '= FAILURES =' pytest.log | return $?
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s): return 0 if match found
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - ! grep -vq '= FAILURES =' pytest.log
+  - ! grep -q '= FAILURES =' pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -106,11 +106,11 @@ script:
 
   # run unit test(s)
   - docker network create -d bridge app_nw
-  - docker run --name base --net=app_nw -d container-default
-  - docker run --name redis --net=app_nw -d container-redis
-  - docker run --name webserver --net=app_nw -d container-webserver run
-  - docker run --name database --net=app_nw -d container-database
-  - docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stdout | grep -q '= FAILURES =' && return 1
+  - docker run --name base --net=app_nw -d container-default \
+    && docker run --name redis --net=app_nw -d container-redis \
+    && docker run --name webserver --net=app_nw -d container-webserver run \
+    && docker run --name database --net=app_nw -d container-database \
+    && docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stdout | grep -q '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,12 +105,12 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && \\
-    docker run --name base --net=app_nw -d container-default && \\
-    docker run --name redis --net=app_nw -d container-redis && \\
-    docker run --name webserver --net=app_nw -d container-webserver run && \\
-    docker run --name database --net=app_nw -d container-database && \\
-    docker run --name webserver-pytest --net=app_nw -it container-webserver test | \\
+  - docker network create -d bridge app_nw && \
+    docker run --name base --net=app_nw -d container-default && \
+    docker run --name redis --net=app_nw -d container-redis && \
+    docker run --name webserver --net=app_nw -d container-webserver run && \
+    docker run --name database --net=app_nw -d container-database && \
+    docker run --name webserver-pytest --net=app_nw -it container-webserver test | \
     grep '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -q '= FAILURES =' && exit 1
+  - cat pytest.log | grep -vq '= FAILURES =' && exit 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,12 +105,12 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw \
-    && docker run --name base --net=app_nw -d container-default \
-    && docker run --name redis --net=app_nw -d container-redis \
-    && docker run --name webserver --net=app_nw -d container-webserver run \
-    && docker run --name database --net=app_nw -d container-database \
-    && docker run --name webserver-pytest --net=app_nw -it container-webserver test | \
+  - docker network create -d bridge app_nw
+  - docker run --name base --net=app_nw -d container-default
+  - docker run --name redis --net=app_nw -d container-redis
+  - docker run --name webserver --net=app_nw -d container-webserver run
+  - docker run --name database --net=app_nw -d container-database
+  - docker run --name webserver-pytest --net=app_nw -it container-webserver test | \
     grep '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s): return 0 if match found
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -vq '= FAILURES =' || return 1
+  - ! grep -vq '= FAILURES =' pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -110,8 +110,7 @@ script:
   - docker run --name redis --net=app_nw -d container-redis
   - docker run --name webserver --net=app_nw -d container-webserver run
   - docker run --name database --net=app_nw -d container-database
-  - docker run --name webserver-pytest --net=app_nw -it container-webserver test | \
-    grep '= FAILURES =' && return 1
+  - docker run --name webserver-pytest --net=app_nw -it container-webserver test | grep '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default docker run --name redis --net=app_nw -d container-redis docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test | grep -q '= FAILURES =' && return 1
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default docker run --name redis --net=app_nw -d container-redis docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stderr | grep -vq '= FAILURES ='
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test > pytest.log | cat pytest.log | grep -q '= FAILURES =' && cat pytest.log
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log | tee pytest.log | grep -q '= FAILURES ='; cat pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,12 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw
-  - docker run --name base --net=app_nw -d container-default \
-    && docker run --name redis --net=app_nw -d container-redis \
-    && docker run --name webserver --net=app_nw -d container-webserver run \
-    && docker run --name database --net=app_nw -d container-database \
-    && docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stdout | grep -q '= FAILURES =' && return 1
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default docker run --name redis --net=app_nw -d container-redis docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stdout | grep -q '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -q '= FAILURES ='
+  - cat pytest.log | grep -q '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -106,11 +106,10 @@ script:
 
   # run unit test(s): return 0 if match found
   #
-  # Note: travis has some issues with the exclamation mark, so subshell is used
+  # Note: travis has issues with the exclamation mark, so subshell is used
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - TEST_OUTPUT=$(! grep -q '= FAILURES =' pytest.log)
-  - echo TEST_OUTPUT
+  - (! grep -q '= FAILURES =' pytest.log)
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - grep -c '= FAILURES =' pytest.log | return $?
+  - cat pytest.log | grep -q '= FAILURES =' && exit 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s): return 0 if match found
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - ! grep -q '= FAILURES =' pytest.log
+  - (! grep -q '= FAILURES =' pytest.log)
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -110,7 +110,9 @@ script:
   - docker run --name redis --net=app_nw -d container-redis
   - docker run --name webserver --net=app_nw -d container-webserver run
   - docker run --name database --net=app_nw -d container-database
-  - docker run --name webserver-pytest --net=app_nw -it container-webserver test | grep '= FAILURES =' && return 1
+  - output = $(docker run --name webserver-pytest --net=app_nw -it container-webserver test)
+  - echo $output
+  - echo $output | grep -q '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -106,7 +106,7 @@ script:
 
   # run unit test(s): return 0 if match found
   #
-  # Note: travis has issues with the exclamation mark, so subshell is used
+  # Note: travis has issues with exclamation, so is contained in a subshell
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
   - (! grep -q '= FAILURES =' pytest.log)

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log | tee pytest.log | grep -q '= FAILURES ='; cat pytest.log
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log | cat pytest.log | grep -q '= FAILURES ='; cat pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,13 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test
+  - docker network create -d bridge app_nw && \\
+    docker run --name base --net=app_nw -d container-default && \\
+    docker run --name redis --net=app_nw -d container-redis && \\
+    docker run --name webserver --net=app_nw -d container-webserver run && \\
+    docker run --name database --net=app_nw -d container-database && \\
+    docker run --name webserver-pytest --net=app_nw -it container-webserver test | \\
+    grep '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -104,10 +104,10 @@ script:
   - docker build -f database.dockerfile -t container-database .
   - docker build -f webserver.dockerfile -t container-webserver .
 
-  # run unit test(s)
+  # run unit test(s): return 0 if match found
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -v '= FAILURES =' && exit 1
+  - cat pytest.log | grep -q '= FAILURES =' || return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default docker run --name redis --net=app_nw -d container-redis docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stderr | grep -vq '= FAILURES ='
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test 2>&1 | tee /dev/stdout | grep -vq '= FAILURES ='
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test 2>&1 | tee /dev/stdout | grep -vq '= FAILURES ='
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test > pytest.log | cat pytest.log | grep -q '= FAILURES =' && cat pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - grep -q '= FAILURES =' pytest.log
+  - grep -vq '= FAILURES =' pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -110,9 +110,7 @@ script:
   - docker run --name redis --net=app_nw -d container-redis
   - docker run --name webserver --net=app_nw -d container-webserver run
   - docker run --name database --net=app_nw -d container-database
-  - output = $(docker run --name webserver-pytest --net=app_nw -it container-webserver test)
-  - echo $output
-  - echo $output | grep -q '= FAILURES =' && return 1
+  - docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stdout | grep -q '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -q '= FAILURES =' && return 1
+  - cat pytest.log | grep -vq '= FAILURES ='
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,7 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log | cat pytest.log | grep -q '= FAILURES ='; cat pytest.log
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log | cat pytest.log | grep -q '= FAILURES =' && return 1; cat pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -vq '= FAILURES =' && exit 1
+  - cat pytest.log | grep -v '= FAILURES =' && exit 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -104,6 +104,13 @@ script:
   - docker build -f database.dockerfile -t container-database .
   - docker build -f webserver.dockerfile -t container-webserver .
 
-  # run unit test(s), list all containers
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test
+  # run unit test(s)
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default docker run --name redis --net=app_nw -d container-redis docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test | grep -q '= FAILURES =' && return 1
+
+  # check exit code: 'docker exec' will fail if the container has an exit code
+  - docker exec -it webserver echo "'docker exec' only works on an active running container"
+  - docker exec -it database echo "'docker exec' only works on an active running container"
+  - docker exec -it redis echo "'docker exec' only works on an active running container"
+
+  # list all containers
   - docker ps -a

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,7 +105,9 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log | cat pytest.log | grep -q '= FAILURES =' && return 1; cat pytest.log
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
+  - cat pytest.log
+  - cat pytest.log | grep -q '= FAILURES =' && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - grep -q '= FAILURES =' pytest.log && return 1
+  - grep -q '= FAILURES =' pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - grep -vq '= FAILURES =' pytest.log
+  - grep -c '= FAILURES =' pytest.log
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s)
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -vq '= FAILURES ='
+  - grep -q '= FAILURES =' pytest.log && return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -105,9 +105,12 @@ script:
   - docker build -f webserver.dockerfile -t container-webserver .
 
   # run unit test(s): return 0 if match found
+  #
+  # Note: travis has some issues with the exclamation mark, so subshell is used
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - (! grep -q '= FAILURES =' pytest.log)
+  - TEST_OUTPUT=$(! grep -q '= FAILURES =' pytest.log)
+  - echo TEST_OUTPUT
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -107,7 +107,7 @@ script:
   # run unit test(s): return 0 if match found
   - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test  > pytest.log
   - cat pytest.log
-  - cat pytest.log | grep -q '= FAILURES =' || return 1
+  - cat pytest.log | grep -vq '= FAILURES =' || return 1
 
   # check exit code: 'docker exec' will fail if the container has an exit code
   - docker exec -it webserver echo "'docker exec' only works on an active running container"

--- a/.travis.yml
+++ b/.travis.yml
@@ -104,13 +104,6 @@ script:
   - docker build -f database.dockerfile -t container-database .
   - docker build -f webserver.dockerfile -t container-webserver .
 
-  # run unit test(s)
-  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default docker run --name redis --net=app_nw -d container-redis docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test | tee /dev/stdout | grep -q '= FAILURES =' && return 1
-
-  # check exit code: 'docker exec' will fail if the container has an exit code
-  - docker exec -it webserver echo "'docker exec' only works on an active running container"
-  - docker exec -it database echo "'docker exec' only works on an active running container"
-  - docker exec -it redis echo "'docker exec' only works on an active running container"
-
-  # list all containers
+  # run unit test(s), list all containers
+  - docker network create -d bridge app_nw && docker run --name base --net=app_nw -d container-default && docker run --name redis --net=app_nw -d container-redis && docker run --name webserver --net=app_nw -d container-webserver run && docker run --name database --net=app_nw -d container-database && docker run --name webserver-pytest --net=app_nw -it container-webserver test
   - docker ps -a

--- a/database.dockerfile
+++ b/database.dockerfile
@@ -6,6 +6,7 @@ ENV ENVIRONMENT docker
 ENV ENVIRONMENT_DIR $ROOT_PROJECT/puppet/environment/$ENVIRONMENT
 
 ## provision with puppet
+RUN cat $ENVIRONMENT_DIR/manifests/setup_database.pp
 RUN /opt/puppetlabs/bin/puppet apply $ENVIRONMENT_DIR/manifests/setup_database.pp --modulepath=$ENVIRONMENT_DIR/modules_contrib:$ENVIRONMENT_DIR/modules --confdir=$ROOT_PROJECT/test
 
 ## executed everytime container starts

--- a/database.dockerfile
+++ b/database.dockerfile
@@ -6,7 +6,6 @@ ENV ENVIRONMENT docker
 ENV ENVIRONMENT_DIR $ROOT_PROJECT/puppet/environment/$ENVIRONMENT
 
 ## provision with puppet
-RUN cat $ENVIRONMENT_DIR/manifests/setup_database.pp
 RUN /opt/puppetlabs/bin/puppet apply $ENVIRONMENT_DIR/manifests/setup_database.pp --modulepath=$ENVIRONMENT_DIR/modules_contrib:$ENVIRONMENT_DIR/modules --confdir=$ROOT_PROJECT/test
 
 ## executed everytime container starts

--- a/hiera/packages.yaml
+++ b/hiera/packages.yaml
@@ -20,7 +20,7 @@ development:
         fetch: '0.11.0'
         uglify-js: '2.6.2'
         imagemin: '5.0.0'
-        node-sass: '3.5.2'
+        node-sass: '3.7.0'
         babel-core: '6.7.4'
         browserify: '13.0.0'
         babelify: '7.2.0'

--- a/puppet/environment/docker/manifests/setup_database.pp
+++ b/puppet/environment/docker/manifests/setup_database.pp
@@ -5,14 +5,24 @@
 ###       https://github.com/jeff1evesque/machine-learning/issues/2349
 ###
 
-## install mariadb
-include database::server
+## install sql
+class install_sql {
+    ## install mariadb
+    include database::server
 
-## install mariadb client
-include database::client
+    ## install mariadb client
+    include database::client
 
-## install mariad bindings
-include database::bindings
+    ## install mariad bindings
+    include database::bindings
+}
 
 ## create database tables
-include database::database
+class create_db {
+    require install_sql
+    include database::database
+}
+
+## initiate
+include create_db
+

--- a/puppet/environment/docker/manifests/setup_database.pp
+++ b/puppet/environment/docker/manifests/setup_database.pp
@@ -20,7 +20,7 @@ class install_sql {
 ## create database tables
 class create_db {
     require install_sql
-    include database::database
+    contain database::database
 }
 
 ## initiate

--- a/puppet/environment/docker/manifests/setup_database.pp
+++ b/puppet/environment/docker/manifests/setup_database.pp
@@ -8,13 +8,13 @@
 ## install sql
 class install_sql {
     ## install mariadb
-    include database::server
+    contain database::server
 
     ## install mariadb client
-    include database::client
+    contain database::client
 
     ## install mariad bindings
-    include database::bindings
+    contain database::bindings
 }
 
 ## create database tables

--- a/puppet/environment/docker/modules/database/manifests/bindings.pp
+++ b/puppet/environment/docker/modules/database/manifests/bindings.pp
@@ -6,4 +6,5 @@ class database::bindings {
     class { '::mysql::bindings':
         python_enable => true,
     }
+    contain mysql::bindings
 }

--- a/puppet/environment/docker/modules/database/manifests/client.pp
+++ b/puppet/environment/docker/modules/database/manifests/client.pp
@@ -7,4 +7,5 @@ class database::client {
     class { '::mysql::client':
         package_name => 'mariadb-client',
     }
+    contain mysql::client
 }

--- a/puppet/environment/docker/modules/database/manifests/server.pp
+++ b/puppet/environment/docker/modules/database/manifests/server.pp
@@ -68,4 +68,5 @@ class database::server {
             }
         },
     }
+    contain mysql::server
 }

--- a/test/programmatic_interface/pytest_svm_session.py
+++ b/test/programmatic_interface/pytest_svm_session.py
@@ -53,7 +53,7 @@ def get_sample_json(jsonfile, model_type):
             json_dataset = json.load(json_file)
 
     except Exception as error:
-        raise IOError(error)
+        pytest.fail(error)
 
     return json.dumps(json_dataset)
 

--- a/test/programmatic_interface/pytest_svm_session.py
+++ b/test/programmatic_interface/pytest_svm_session.py
@@ -75,7 +75,7 @@ def test_data_new(client, live_server):
     res = client.post(
         get_endpoint(),
         headers={'Content-Type': 'application/json'},
-        data=get_sample_json('svm-data-new.json', 'svm')
+        data=get_sample_json('svm-data-new-typo.json', 'svm')
     )
 
     assert res.status_code == 200

--- a/test/programmatic_interface/pytest_svm_session.py
+++ b/test/programmatic_interface/pytest_svm_session.py
@@ -75,7 +75,7 @@ def test_data_new(client, live_server):
     res = client.post(
         get_endpoint(),
         headers={'Content-Type': 'application/json'},
-        data=get_sample_json('svm-data-new-typo.json', 'svm')
+        data=get_sample_json('svm-data-new.json', 'svm')
     )
 
     assert res.status_code == 200

--- a/test/programmatic_interface/pytest_svm_session.py
+++ b/test/programmatic_interface/pytest_svm_session.py
@@ -35,20 +35,26 @@ def get_sample_json(jsonfile, model_type):
 
     # open file
     json_dataset = None
-    with open(
-        os.path.join(
-            root,
-            'interface',
-            'static',
-            'data',
-            'json',
-            'programmatic_interface',
-            model_type,
-            jsonfile
-        ),
-        'r'
-    ) as json_file:
-        json_dataset = json.load(json_file)
+
+    try:
+        with open(
+            os.path.join(
+                root,
+                'interface',
+                'static',
+                'data',
+                'json',
+                'programmatic_interface',
+                model_type,
+                jsonfile
+            ),
+            'r'
+        ) as json_file:
+            json_dataset = json.load(json_file)
+
+    except Exception as error:
+        raise IOError(error)
+
     return json.dumps(json_dataset)
 
 
@@ -68,7 +74,7 @@ def test_data_new(client, live_server):
     res = client.post(
         get_endpoint(),
         headers={'Content-Type': 'application/json'},
-        data=get_sample_json('svm-data-new.json', 'svm')
+        data=get_sample_json('svm-data-new-typo.json', 'svm')
     )
 
     assert res.status_code == 200

--- a/test/programmatic_interface/pytest_svm_session.py
+++ b/test/programmatic_interface/pytest_svm_session.py
@@ -18,6 +18,7 @@ Note: the 'pytest' instances can further be reviewed:
 '''
 
 import json
+import pytest
 import os.path
 from flask import url_for
 from flask import current_app

--- a/test/programmatic_interface/pytest_svr_session.py
+++ b/test/programmatic_interface/pytest_svr_session.py
@@ -53,7 +53,7 @@ def get_sample_json(jsonfile, model_type):
             json_dataset = json.load(json_file)
 
     except Exception as error:
-        raise IOError(error)
+        pytest.fail(error)
 
     return json.dumps(json_dataset)
 

--- a/test/programmatic_interface/pytest_svr_session.py
+++ b/test/programmatic_interface/pytest_svr_session.py
@@ -18,6 +18,7 @@ Note: the 'pytest' instances can further be reviewed:
 '''
 
 import json
+import pytest
 import os.path
 from flask import url_for
 from flask import current_app

--- a/test/programmatic_interface/pytest_svr_session.py
+++ b/test/programmatic_interface/pytest_svr_session.py
@@ -35,20 +35,26 @@ def get_sample_json(jsonfile, model_type):
 
     # open file
     json_dataset = None
-    with open(
-        os.path.join(
-            root,
-            'interface',
-            'static',
-            'data',
-            'json',
-            'programmatic_interface',
-            model_type,
-            jsonfile
-        ),
-        'r'
-    ) as json_file:
-        json_dataset = json.load(json_file)
+
+    try:
+        with open(
+            os.path.join(
+                root,
+                'interface',
+                'static',
+                'data',
+                'json',
+                'programmatic_interface',
+                model_type,
+                jsonfile
+            ),
+            'r'
+        ) as json_file:
+            json_dataset = json.load(json_file)
+
+    except Exception as error:
+        raise IOError(error)
+
     return json.dumps(json_dataset)
 
 

